patches/server/0957-Add-Listing-API-for-Player.patch

```diff
@@ -0,0 +1,168 @@
+-        return new ClientboundPlayerInfoUpdatePacket(enumSet, players);
++        // return new ClientboundPlayerInfoUpdatePacket(enumSet, players); // Paper
++        // Paper start
++        List<ClientboundPlayerInfoUpdatePacket.Entry> entries = new java.util.ArrayList<>(players.size());
++        for (ServerPlayer player : players) {
++            entries.add(new ClientboundPlayerInfoUpdatePacket.Entry(player, forPlayer.getBukkitEntity().isListed(player.getBukkitEntity())));
++        }
++        return new ClientboundPlayerInfoUpdatePacket(enumSet, entries);
++        // Paper end
++    }
++    // Paper start
++    public static ClientboundPlayerInfoUpdatePacket updateListed(UUID playerInfoId, boolean listed) {
++        EnumSet<ClientboundPlayerInfoUpdatePacket.Action> enumSet = EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LISTED);
++        return new ClientboundPlayerInfoUpdatePacket(enumSet, new ClientboundPlayerInfoUpdatePacket.Entry(playerInfoId, listed));
+     }
++    // Paper end
+ 
+     public ClientboundPlayerInfoUpdatePacket(FriendlyByteBuf buf) {
+         this.actions = buf.readEnumSet(ClientboundPlayerInfoUpdatePacket.Action.class);
+@@ -138,6 +162,15 @@ public class ClientboundPlayerInfoUpdatePacket implements Packet<ClientGamePacke
+         Entry(ServerPlayer player) {
+             this(player.getUUID(), player.getGameProfile(), true, player.latency, player.gameMode.getGameModeForPlayer(), player.getTabListDisplayName(), Util.mapNullable(player.getChatSession(), RemoteChatSession::asData));
+         }
++        // Paper start
++        Entry(ServerPlayer player, boolean listed) {
++            this(player.getUUID(), player.getGameProfile(), listed, player.latency, player.gameMode.getGameModeForPlayer(), player.getTabListDisplayName(), Util.mapNullable(player.getChatSession(), RemoteChatSession::asData));
++        }
++
++        Entry(UUID profileId, boolean listed) {
++            this(profileId, null, listed, 0, GameType.DEFAULT_MODE, null, null);
++        }
++        // Paper end
+     }
+ 
+     static class EntryBuilder {
+diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
+index 4b754f6eae683248d7fe11d6d6cb168d5dd696a2..23a6f73a724ec6b9288f681112e5185d23f8ea35 100644
+--- a/src/main/java/net/minecraft/server/players/PlayerList.java
++++ b/src/main/java/net/minecraft/server/players/PlayerList.java
+@@ -331,7 +331,7 @@ public abstract class PlayerList {
+         // CraftBukkit end
+ 
+         // CraftBukkit start - sendAll above replaced with this loop
+-        ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(player));
++        ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(player), player); // Paper
```

Comment:
This doesn't look right, you need to check visibility for other players, not the player itself